File: src/data/object/validator.js
const Ajv = require('ajv');
const ajv = new Ajv();

class Validator {
  /**
   * validates value against constraints
   * @param {Object} value
   * @param {Object} [constraints]
   * @return {Array<Error>}
   */
  validate(value,
           constraints) {
    if(!constraints){
      return [];
    }

    ajv.validate(constraints, value);
    return ajv.errors;
  }
}

// export as singleton
module.exports = new Validator();

File: sdks/go/opspec/interpreter/call/op/params/param/file/validator.go
package file

import (
	"errors"
	"fmt"
	"github.com/golang-interfaces/ios"
	"github.com/opctl/opctl/sdks/go/model"
)

type Validator interface {
	Validate(
		value *model.Value,
	) []error
}

func NewValidator() Validator {
	return _validator{
		os: ios.New(),
	}
}

type _validator struct {
	os ios.IOS
}

// Validate validates a value against a string parameter
func (vdt _validator) Validate(
	value *model.Value,
) []error {
	if nil == value || nil == value.File {
		return []error{errors.New("file required")}
	}

	fileInfo, err := vdt.os.Stat(*value.File)
	if nil != err {
		return []error{err}
	} else if !fileInfo.Mode().IsRegular() {
		return []error{fmt.Errorf("%v not a file", *value.File)}
	}
	return []error{}
}
